ts/lib/promise.ts

```diff
@@ -0,0 +1,13 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+/* eslint
+@typescript-eslint/no-non-null-assertion: "off",
+ */
+
+export function promiseWithResolver<T>(): [Promise<T>, (value: T) => void] {
```

Comment:
Not a request for the current PR, but just a general comment - it would be nice to see a few more docstrings in future changes, especially on the more abstract/complicated parts. I'm guilty of not documenting my code as much as I should as well, so consider it advice to the both of us :-)